test/extensions/filters/http/geoip/geoip_filter_test.cc

```diff
@@ -0,0 +1,278 @@
+                    const std::string& geo_header_name, const std::string& value) {
+    EXPECT_TRUE(request_headers.has(geo_header_name));
+    EXPECT_EQ(
+        value,
+        request_headers.get(Http::LowerCaseString(geo_header_name))[0]->value().getStringView());
+  }
+
+  ~GeoipFilterTest() override { filter_->onDestroy(); }
+
+  NiceMock<Stats::MockStore> stats_;
+  GeoipFilterConfigSharedPtr config_;
+  std::unique_ptr<GeoipFilter> filter_;
+  std::unique_ptr<DummyGeoipProviderFactory> dummy_factory_;
+  MockDriverSharedPtr dummy_driver_;
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> filter_callbacks_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  absl::optional<std::string> empty_response_;
+  absl::optional<std::string> dummy_city_;
+  absl::optional<std::string> dummy_country_;
+  absl::optional<std::string> dummy_region_;
+  absl::optional<std::string> dummy_asn_;
+  absl::optional<std::string> dummy_anon_response_;
+  Network::Address::InstanceConstSharedPtr captured_address_;
+};
+
+TEST_F(GeoipFilterTest, NoXffSuccessfulLookup) {
+  initializeProviderFactory();
+  const std::string external_request_yaml = R"EOF(
+    geo_headers_to_add:
+      city: "x-geo-city"
+    provider:
+        name: "envoy.geoip_providers.dummy"
+)EOF";
+  initializeFilter(external_request_yaml);
+  Http::TestRequestHeaderMapImpl request_headers;
+  expectStats("x-geo-city");
+  Network::Address::InstanceConstSharedPtr remote_address =
+      Network::Utility::parseInternetAddress("1.2.3.4");
+  filter_callbacks_.stream_info_.downstream_connection_info_provider_->setRemoteAddress(
+      remote_address);
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_->decodeHeaders(request_headers, false));
+  EXPECT_CALL(filter_callbacks_, continueDecoding());
+  filter_->onLookupComplete(dummy_city_, "x-geo-city");
+  EXPECT_EQ(1, request_headers.size());
+  expectHeader(request_headers, "x-geo-city", "dummy_city");
+  EXPECT_EQ("1.2.3.4:0", captured_address_->asString());
+}
+
+TEST_F(GeoipFilterTest, UseXffSuccessfulLookup) {
+  initializeProviderFactory();
+  const std::string external_request_yaml = R"EOF(
+    use_xff: true
+    xff_num_trusted_hops: 1
+    geo_headers_to_add:
+      region: "x-geo-region"
+    provider:
+        name: "envoy.geoip_providers.dummy"
+)EOF";
+  initializeFilter(external_request_yaml);
+  Http::TestRequestHeaderMapImpl request_headers;
+  request_headers.addCopy("x-forwarded-for", "10.0.0.1,10.0.0.2");
+  expectStats("x-geo-region");
+  Network::Address::InstanceConstSharedPtr remote_address =
+      Network::Utility::parseInternetAddress("1.2.3.4");
+  filter_callbacks_.stream_info_.downstream_connection_info_provider_->setRemoteAddress(
+      remote_address);
+  EXPECT_EQ(Http::FilterHeadersStatus::StopIteration,
+            filter_->decodeHeaders(request_headers, false));
+  EXPECT_CALL(filter_callbacks_, continueDecoding());
+  filter_->onLookupComplete(dummy_region_, "x-geo-region");
```

Comment:
Could you make dummy_driver_ also SaveArg<2> (or 3), the callback, and then instead of directly calling onLookupComplete, call it through the captured callback? That would get you better coverage of how the filter actually works. Current testing doesn't verify that the driver gets a correct callback, nor that calling the driver's callback does anything.

(Might make more sense to not use SaveArg, and instead make a custom function that captures all the args into a specified dummy_driver_ member variable struct, so you can save the different callbacks to different slots without having to copy-paste 8 near-duplicate sets of SaveArg.)